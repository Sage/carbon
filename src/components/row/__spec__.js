import React from 'react';
import Immutable from 'immutable';
import TestUtils from 'react/lib/ReactTestUtils';
import Row from './row';

describe('Row', () => {

  describe('when column number is NOT passed', () => {
    let instance;
    let columns;

    beforeEach(() => {
      instance = TestUtils.renderIntoDocument(
        <Row className="foobar">
          <div columnOffset={2}>Foo</div>
          <div columnSpan={3}>Bar</div>
          { null }
          <div columnClasses='extra-class'>Bar</div>
          <div columnAlign='center'>Bar</div>
        </Row>
      );

      columns = TestUtils.scryRenderedDOMComponentsWithClass(instance, 'carbon-row__column');
    });

    describe('render', () => {
      it('renders a parent div with calculated CSS classes', () => {
        let rowNode = TestUtils.findRenderedDOMComponentWithClass(instance, 'carbon-row')
        expect(rowNode.className).toEqual('carbon-row foobar carbon-row--columns-4');
      });

      it('renders the correct amount of columns', () => {
        expect(columns.length).toEqual(4);
      });
    });

    describe('with immutable data', () => {
      it('renders the correct number of columns', () => {
        let data = Immutable.fromJS([{
          name: 'foo'
        }, {
          name: 'bar'
        }]);

        instance = TestUtils.renderIntoDocument(
          <Row className="foobar">
            {
              data.map((item) => {
                return <div>{ item.get('name') }</div>;
              })
            }
          </Row>
        );

        columns = TestUtils.scryRenderedDOMComponentsWithClass(instance, 'carbon-row__column');
        expect(columns.length).toEqual(2);
      });
    });

    describe('with no children', () => {
      it('allows render', () => {
        instance = TestUtils.renderIntoDocument(
          <Row>{ null }</Row>
        );
        let rowNode = TestUtils.findRenderedDOMComponentWithClass(instance, 'carbon-row')
        expect(rowNode).toBeTruthy();
      });
    });

    describe('with no children in an array', () => {
      it('allows render', () => {
        let children = [];
        instance = TestUtils.renderIntoDocument(
          <Row>{ children }</Row>
        );
        let rowNode = TestUtils.findRenderedDOMComponentWithClass(instance, 'carbon-row')
        expect(rowNode).toBeTruthy();
      });
    });

    describe('Column offset', () => {
      it('renders a div with an additional offset CSS class', () => {
        expect(columns[0].className).toEqual('carbon-row__column carbon-row__column--offset-2');
      });
    });

    describe('Column span', () => {
      it('renders a div with an additional span CSS class', () => {
        expect(columns[1].className).toEqual('carbon-row__column carbon-row__column--span-3');
      });
    });

    describe('Column classes', () => {
      it('renders a div with all additional column classes', () => {
        expect(columns[2].className).toEqual('carbon-row__column extra-class');
      });
    });

    describe('Column align', () => {
      it('renders a div with alignment class', () => {
        expect(columns[3].className).toEqual('carbon-row__column carbon-row__column--align-center');
      });
    });
  });

  describe('When column number is passed', () => {
    let instance;

    beforeEach(() => {
      instance = TestUtils.renderIntoDocument(
        <Row columns={2}>
          <div>Foo</div>
          <div>Bar</div>
        </Row>
      );
    });

    it('renders a parent div with calculated CSS classes', () => {
      let rowNode = TestUtils.findRenderedDOMComponentWithClass(instance, 'carbon-row')
      expect(rowNode.className).toEqual('carbon-row carbon-row--columns-2');
    });
  });

  describe('When there is only 1 column', () => {
    let instance;

    beforeEach(() => {
      instance = TestUtils.renderIntoDocument(
        <Row>
          <div>Foo</div>
        </Row>
      );
    });

    it('renders a parent div with calculated CSS classes', () => {
      let rowNode = TestUtils.findRenderedDOMComponentWithClass(instance, 'carbon-row')
      expect(rowNode.className).toEqual('carbon-row carbon-row--columns-1');
    });
  });
});
