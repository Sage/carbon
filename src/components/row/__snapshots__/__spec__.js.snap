// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row columnClasses passes the prop to its children 1`] = `
<div
  className="carbon-row carbon-row--gutter-medium carbon-row--columns-1"
  data-component="row"
>
  <Column
    columnAlign="left"
    columnClasses="myChildClasses"
    columnOffset="0"
    columnSpan="1"
    key=".0"
  />
</div>
`;

exports[`Row columnDivide when columnDivide is true passes columnDivide to its children 1`] = `
<div
  className="carbon-row carbon-row--gutter-medium carbon-row--columns-1"
  data-component="row"
>
  <Column
    columnAlign="left"
    columnDivide={true}
    columnOffset="0"
    columnSpan="1"
    key=".0"
  />
</div>
`;

exports[`Row columns when a null column is passed ignores the null child 1`] = `
<div
  className="carbon-row carbon-row--gutter-medium carbon-row--columns-2"
  data-component="row"
>
  <Column
    columnAlign="left"
    columnOffset="0"
    columnSpan="1"
    key=".$0"
  />
  <Column
    columnAlign="left"
    columnOffset="0"
    columnSpan="1"
    key=".$1"
  />
</div>
`;

exports[`Row columns when column number is not passed determines the columns from the number of children 1`] = `
<div
  className="carbon-row carbon-row--gutter-medium carbon-row--columns-4"
  data-component="row"
>
  <Column
    columnAlign="left"
    columnOffset="0"
    columnSpan="1"
    key=".0"
  />
  <Column
    columnAlign="left"
    columnOffset="0"
    columnSpan="1"
    key=".1"
  />
  <Column
    columnAlign="left"
    columnOffset="0"
    columnSpan="1"
    key=".2"
  />
  <Column
    columnAlign="left"
    columnOffset="0"
    columnSpan="1"
    key=".3"
  />
</div>
`;

exports[`Row columns when column number is passed sets the passed column as the number of columns 1`] = `
<div
  className="carbon-row carbon-row--gutter-medium carbon-row--columns-50"
  data-component="row"
>
  <Column
    columnAlign="left"
    columnOffset="0"
    columnSpan="1"
    key=".0"
  />
</div>
`;

exports[`Row gutter applies a gutter class 1`] = `
<div
  className="carbon-row carbon-row--gutter-small carbon-row--columns-0"
  data-component="row"
/>
`;

exports[`Row render renders a parent div with calculated CSS classes 1`] = `
<div
  className="carbon-row carbon-row--gutter-medium carbon-row--columns-0"
  data-component="row"
/>
`;
